### ISSUE TITLE
Library used for nullable, etc. annotations

### ISSUE THREAD

**Issue by zeroeightysix** [DaemonicLabs/fiber#30]
In https://github.com/FabricMC/fabric/issues/499 the library used for the `@Deprecated` is being discussed. I'd like to use the same library for the `@Nullable` annotations (and variants).

Note IDEA currently provides the following warning for all usages of `javax` annotations:
```
Not 'javax.annotation.Nullable' but 'org.jetbrains.annotations.Nullable' would be used for code generation. 
```

This issue serves as a tracker until https://github.com/FabricMC/fabric/issues/499 is solved.

**Comment by liach** [comment]
Should we use jetbrains or javax? javax annotations imo are not sustainable in long term as we might have things like nested nullable types

**Comment by zeroeightysix** [comment]
I prefer jetbrains annotations as well. Fiber currently uses javax solely because it's in the standard lib.

**Comment by liach** [comment]
With jetbrain annotations/other type annotations and some tweaks to https://github.com/FabLabsMC/fiber/blob/9096f8160b17a882bc3bab3ac7ba0c822dff6948/src/main/java/me/zeroeightsix/fiber/annotation/magic/TypeMagic.java#L34, it will be able to handle the nullability of nested types properly

**Comment by Pyrofab** [comment]
I've been dying to use Jetbrains annotations everywhere in my PRs. With that said, would Fabric Loader accept a compile-time dependency on those ?

**Comment by liach** [comment]
Yep, but a problem with jetbrain annotations is that they are class-level retention (i.e. aren't exposed at runtime so config's type resolution cannot perform nullable checks unless using asm to grab those annotations)

**Comment by i509VCB** [comment]
Does checker preserve annotations into runtime?

**Comment by liach** [comment]
> 
> 
> Does checker preserve annotations into runtime?

https://github.com/typetools/checker-framework/blob/6cf4e4f2cb829bb39a3bc33f3a11b547f3afe3dd/checker/src/main/java/org/checkerframework/checker/nullness/qual/Nullable.java#L27
It does

**Comment by Pyrofab** [comment]
I did not consider using nullness annotations for runtime checks. That is an excellent idea, it could use a new `NonNullConstraint`.

**Comment by alexis-evelyn** [comment]
I'd vote for Jetbrains. The Javax annotations cause build problems when I'm using CurseGradle on Github Actions (and the error log leads to this project). I've had to manually add the Javax annotations as an implementation even though I personally use Jetbrains.

**Comment by RemainingToast** [comment]
Jetbrains ftw
